homeassistant/components/mqtt/device_trigger.py

```diff
@@ -0,0 +1,277 @@
+
+from homeassistant.components import mqtt
+from homeassistant.components.automation import AutomationActionType
+import homeassistant.components.automation.mqtt as automation_mqtt
+from homeassistant.components.device_automation import TRIGGER_BASE_SCHEMA
+from homeassistant.const import CONF_DEVICE_ID, CONF_DOMAIN, CONF_PLATFORM, CONF_TYPE
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from . import (
+    ATTR_DISCOVERY_HASH,
+    CONF_CONNECTIONS,
+    CONF_DEVICE,
+    CONF_IDENTIFIERS,
+    CONF_PAYLOAD,
+    CONF_QOS,
+    DOMAIN,
+)
+from .discovery import MQTT_DISCOVERY_UPDATED, clear_discovery_hash
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_AUTOMATION_TYPE = "automation_type"
+CONF_DISCOVERY_ID = "discovery_id"
+CONF_SUBTYPE = "subtype"
+CONF_TOPIC = "topic"
+DEFAULT_ENCODING = "utf-8"
+DEVICE = "device"
+
+MQTT_TRIGGER_BASE = {
+    # Trigger when MQTT message is received
+    CONF_PLATFORM: DEVICE,
+    CONF_DOMAIN: DOMAIN,
+}
+
+TRIGGER_SCHEMA = TRIGGER_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_PLATFORM): DEVICE,
+        vol.Required(CONF_DOMAIN): DOMAIN,
+        vol.Required(CONF_DEVICE_ID): str,
+        vol.Required(CONF_DISCOVERY_ID): str,
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    }
+)
+
+TRIGGER_DISCOVERY_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_AUTOMATION_TYPE): str,
+        vol.Required(CONF_DEVICE): mqtt.MQTT_ENTITY_DEVICE_INFO_SCHEMA,
+        vol.Required(CONF_TOPIC): mqtt.valid_subscribe_topic,
+        vol.Optional(CONF_PAYLOAD, default=None): vol.Any(None, cv.string),
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    },
+    mqtt.validate_device_has_at_least_one_identifier,
+)
+
+DEVICE_TRIGGERS = "mqtt_device_triggers"
+ATTACHED_DEVICE_TRIGGERS = "mqtt_attached_device_triggers"
+
+
+@attr.s(slots=True)
+class Trigger:
+    """Device trigger settings."""
+
+    device_id = attr.ib(type=str)
+    discovery_id = attr.ib(type=str)
+    hass = attr.ib(type=HomeAssistantType)
+    payload = attr.ib(type=str)
+    qos = attr.ib(type=int)
+    subtype = attr.ib(type=str)
+    topic = attr.ib(type=str)
+    type = attr.ib(type=str)
+
+    async def attach_trigger(self, action, automation_info):
+        """Attach MQTT trigger."""
+        mqtt_config = {
+            automation_mqtt.CONF_TOPIC: self.topic,
+            automation_mqtt.CONF_ENCODING: DEFAULT_ENCODING,
+            automation_mqtt.CONF_QOS: self.qos,
+        }
+        if self.payload:
+            mqtt_config[CONF_PAYLOAD] = self.payload
+
+        return await automation_mqtt.async_attach_trigger(
```

Comment:
Shouldn't you store the attached trigger in this class?

Reply:
OK, refactored accordingly.